src/filter.jl

```diff
@@ -85,3 +85,206 @@ function hamilton_filter(y::AbstractVector, h::Integer)
     y_trend = y - y_cycle
     return y_cycle, y_trend
 end
+
+"""
+- `g::Function`: Parameterized `Function` that recieves three arguments 
+                 (data at each period, state at each period, parameter) and return the density.
+                 For example, if the data `x <: Real` is generated by `N(mu_s, sigma_s)` where mean `mu` and std `sigma` is
+                 state dependent, `g` should be defined as 
+                 ```
+                 g(x, s, parameter) = exp(-((x-parameter.mu[s])^2)/(2*parameter.sigma[s]^2))/sqrt(2*pi*parameter.sigma[s]^2)
+                 ```
+- `y::AbstractArray`: Collection of data whose row corresponds to periods
+- `parameter`: Collection of parameters in the model. Each parameter must be a vector that stores the values for each state.
+               For example, if you want to set `mu` at state 1 is 0 and `mu` at state 2 is 0.5 and 
+               `sigma` at state 1 is 3 and `sigma` at state 2 is 4, `parameter` should be
+               ```
+               parameter = (mu = [0, 0.5], sigma = [3, 4])
+               ```
+               Alternatively, you can define your own type and use it:
+               ```
+               struct ParaType
+                   mu::Vector
+                   sigma::Vector
+               end
+               parameter = ParaType([0, 0.5], [3, 4])
+               ```
+- `P::AbstractMatrix`: Transition matrix of state. It must be square.
+- `M::Integer`: Number of states. It must be same as the number of rows and columns of `P`. If `M` is not passed, 
+                number of rows of `P` is set as `M`.
+"""
+mutable struct RegimeSwitchingModel{TD <: AbstractArray, TP <: AbstractMatrix, TPS <: AbstractMatrix}
+    g::Function
+    y::TD
+    parameter

```

Comment:
Is there a reason that we don't want this typed? By having a mutable structure and not having this argument typed, we can change it to anything we want -- For example, I can set `rsm.parameter = 1` and it won't complain. Is there a case where this changes types?

Reply:
> Is there a reason that we don't want this typed?

Kind of, I thought some may want to use user-defined `struct` like [this](https://github.com/QuantEcon/QuantEcon.jl/pull/250/files/6fce054b38ad61c04ac13229a61f6a0707740a88#diff-e5f041a6f2fa9d7548572012875b7606R104) rather than `NamedTuple`. But I also think it should be typed. Maybe I can force it to be `NamedTuple`.

Reply:
Perfect -- This is a use case Julia should handle very cleanly! You can add a type parameter to `RegimeSwitchingModel` and then just refer to that type parameter when typing the argument. I'd make suggestion, but I can only figure out how to suggest one line change, but here's what it would look like:

```julia
mutable struct RegimeSwitchingModel{TD <: AbstractArray, TP <: AbstractMatrix, TPS <: AbstractMatrix, TPRM}
g::Function
y::TD
parameter::TPRM
...
```

This will allow the functions to specialize to the user's type (there will be a new type of  `RegimeSwitchingModel` for every type that gets put in as a parameter) and still prevent people from doing "weird" changes to the `parameter` data.

Reply:
Oh I see, that makes sense and is way better. Thank you!!